site/app/libraries/NotebookUtils.php

```diff
@@ -5,90 +5,254 @@
-                                'image/jpeg',
-                                'image/gif',
-                                'image/bmp',
-                                'text/plain', // Fall back to text/plain if it is available.
+                        else {
+                            $code_cell[] = [
+                                'type' => 'image',
+                                'image' => "data:" . $output_type . ';base64, ' . $img,
+                                'width' => 0,
+                                'height' => 0,
+                                'alt_text' => $text,
                             ];
-
-                            $output_type = null;
-                            foreach ($mime_types as $mime_type) {
-                                if (isset($output['data'][$mime_type])) {
-                                    $output_type = $mime_type;
-                                    break;
-                                }
-                            }
-
-                            if ($output_type === 'text/plain') {
-                                // Display output text if we don't know how to render the content otherwise
-                                $cells[] = [
-                                    'type' => 'output',
-                                    'output_text' => $output['data']['text/plain'],
-                                ];
-                            }
-                            elseif ($output_type !== null) {
-                                $cells[] = [
-                                    'type' => 'image',
-                                    'image' => 'data:' . $output_type . ';base64, ' . $output['data'][$output_type],
-                                    'width' => 0,
-                                    'height' => 0,
-                                    'alt_text' => $output['data']['text/plain'] ?? '',
-                                ];
-                            }
                         }
                     }
-
                     break;
                 default:
                     break;
             }
         }
 
-        return $cells;
+        return [
+            'cells' => $code_cell,
+            'skipped_output' => $skipped_output
+        ];
+    }
+
+    /**
+     * Truncate text to the defined limit and append a message if truncated.
+     *
+     * @param string|string[] $text
+     * @return array{
+     *     text: string,
+     *     was_truncated: int
+     * }
+     */
+    private static function truncateText(string|array $text): array {
+        $output_text = is_array($text) ? implode($text) : $text;
+        $truncated_text = $output_text;
+        $was_truncated = false;
+
+        // Truncate by line limit if necessary
+        $lines = explode("\n", $truncated_text);
+        if (count($lines) > self::LINE_LIMIT) {
+            $lines = array_slice($lines, 0, self::LINE_LIMIT);
+            $truncated_text = implode("\n", $lines);
```

Comment:
Instead of trying to truncate this by splitting into an array and imploding, you could perform string operations instead.  i.e., find the index of the nth line break character, and then take the substring before that point.  An extra benefit of that approach is that you can simplify the truncation process to instead take the substring from 0 to `max(self::TEXT_LIMIT, position_of_line_limit_char)`.